Variable thread count for multi-threaded GEMMs

timmoon10 opened the pull request:
When performing smallish GEMMs (M\~1000, N\~100, K\~100) on a 36-core system, I've found that I can achieve a significant speedup by reducing the number of OpenMP threads. It turns out that multi-threading is disabled if `(m < nthreads * SWITCH_RATIO) || (n < nthreads * SWITCH_RATIO)`, so we maintain parallel execution with no algorithmic changes if `nthreads` is changed to `min(m / SWITCH_RATIO, n / SWITCH_RATIO, nthreads)`. This pull request makes this change automatically and also clears some unused code that dates back to GotoBLAS.

Commit:
Reducing threads for multi-threaded GEMMs on small matrices.

Commit:
Merge https://github.com/timmoon10/OpenBLAS into develop

martin-frbg commented:
Looks good to me, but could you please provide some performance figures just to illustrate how significant "significant" is ? (And if I read your patch correctly, it should not cause a slowdown on less well endowed systems ?)

Commit:
Increasing flexibility of GEMM benchmark.

m, n, and k can be set to arbitrary constants. A and B matrices can be transposed independently.

timmoon10 commented:
Here are some scaling studies on a 36-core system (2 x Intel 18-core Xeon E5-2695 v4, gcc 4.9.3, TARGET=HASWELL USE_OPENMP=1 NUM_THREADS=36 INTERFACE64=0):

![surface_gpgpu_sgemm](https://user-images.githubusercontent.com/4406448/30991230-1174b93e-a459-11e7-9b7d-130f8a0fdd88.png)

![surface_gpgpu_dgemm](https://user-images.githubusercontent.com/4406448/30991325-81e1790a-a459-11e7-9bda-627182e4369c.png)

![surface_gpgpu_cgemm](https://user-images.githubusercontent.com/4406448/30991359-9d7e2956-a459-11e7-85bd-62b150ccf0a0.png)

![surface_gpgpu_zgemm](https://user-images.githubusercontent.com/4406448/30991342-8f757e9a-a459-11e7-87ae-7989e114f1a7.png)

timmoon10 commented:
Here are some experiments on an 8-core system (Intel Core i7, default build for OSX):

![osx_sgemm](https://user-images.githubusercontent.com/4406448/30991454-f79000cc-a459-11e7-8bc8-db0d4169e4f9.png)

![osx_dgemm](https://user-images.githubusercontent.com/4406448/30991457-f9c3e7c8-a459-11e7-9a88-45bedda75d22.png)

![osx_cgemm](https://user-images.githubusercontent.com/4406448/30991459-fd4f80a0-a459-11e7-8ac2-236dbb93c520.png)

![osx_zgemm](https://user-images.githubusercontent.com/4406448/30991464-01bca352-a45a-11e7-8fc0-6f348f6aec11.png)

brada4 commented:
There is another threading threshold in interface/gemm.c
Graph going down probably means move between L3 of different cores.

brada4 commented:
I forgot to mention that you are on right path, fixed threading threshold as currently is great impairment to say OpenCV/SGEMM
